### PR TITLE
Daemonize Slaves

### PR DESCRIPTION
**Pull request by Hornwitser**
Implementation of #201.  Also implements #208.

The configuration is half baked, I wanted to implement #200 before this but it turned out to be too difficult and/or too much wasted work without doing this first, but this would have been much better with the new config system.

The plugins on the slaves don't work any more as the plugin system is not able to handle multiple instances and that's not really fixable now. We need #220 to get somewhere, but that sort of depends on this being implemented first.  As a result the master branch is almost completely useless after this.

---

I do think this needs a little more polish before being merged, some more unit tests and documentation, but overall this is what I've cooked up the last weeks.  There are also some unrelated changes in here, but they all sort of depend on other changes here.  This is all just a mess.

### CODE REVIEW

**Comment by Hornwitser** [comment]
I have restructured the way things work a little.

- lib/link/requests was renamed to lib/link/messages.
- moved lib/link/commands into clusterctl.  It didn't make sense to have it in lib/link.
- Instead of defining the handlers in the Request definition they are instead defined as methods on the Link sub class.  So the listInstances request handler is now defined in ControlConnection.listInstancesRequestHandler in master.js.  This fixes the problem I was having about these handlers needing to reference things in their environment, and separates lib/link from implementation of the logic on the various sides.
- Introduced an Event class for single direction update like messages.  They are sent by one party and received by the other (I'm not sure if the handlers should be async or not though).  It's basically socket.io' ordinary events except the format of the payload is checked.

**Comment by Hornwitser** [comment]
I've added forwarding of the Factorio output from the instances to the clusterctl tool.  It's somewhat primitive for now.  I also moved out the attaching of message types from the Link class to make it easier to test it.  Other than that there's more tests, documentation and polishing.

**Comment by Hornwitser** [comment]
This is now ready for review.  There's no tests for the master/client/clusterctl code that talks to each other.  I guess it's up to you to decide if this should be added by this PR or not.

**Comment by psihius** [comment]
Well, usually tests should go in the same PR as the features/changes, otherwise, they just may get forgotten and then other changes come in and so on.

I would prefer tests to be here.

File: client.js [hunk 1]
```diff
@@ -52,6 +235,178 @@ class Instance {
...
+					instanceConfig = JSON.parse(await fs.readFile(configPath));
+				} catch (err) {
+					if (err.code === "ENOENT") {
+						continue; // Ignore folders without config.json
+					}
+
+					console.error(`Error occured while parsing ${configPath}: ${err}`);
+				}
+
+				// XXX should probably validate the entire config with a JSON Schema.
```
Comment (psihius) [inline client.js]: Not a bad idea, I will add an issue

File: clusterctl.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	// Try looking in the default master config location
+	try {
+		let masterConfig = JSON.parse(await fs.readFile("config.json"));
+		let token = jwt.sign({ id: "api" }, masterConfig.masterAuthSecret, { expiresIn: 600 });
+		
+		let url;
+		if (masterConfig.sslPort) {
+			url = `https://localhost:${masterConfig.sslPort}`;
+		} else {
+			url = `http://localhost:${masterConfig.masterPort}`;
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply clusterctl.js]: The assumption here is that if the master config exists then the code is being runned on the master server itself.  So localhost is the right address to connect to.  If you create a config for clusterctl with clusterctl create-config this credential discovery will by bypassed.

**Comment by Hornwitser** [comment]
The code is not trivial to test.  It's tightly coupled to the socket.io connection, the master code and the slave/clusterctl code.

**Comment by psihius** [comment]
@Hornwitser time to mock socket lib for testing? :D

**Pull request merged**